pkg/scalers/metrics_api_scaler.go

```diff
@@ -417,8 +464,140 @@ func getValueFromYAMLResponse(body []byte, valueLocation string) (float64, error
+	// parse service name from s.meta.url
+	url, err := neturl.Parse(serviceURL)
+	if err != nil {
+		s.logger.Error(err, "Failed parsing url for metrics API")
+	} else {
+		splittedHost := strings.Split(url.Host, ".")
+		if len(splittedHost) < 2 {
+			return nil, fmt.Errorf("invalid hostname %s : expected at least 2 elements, first being service name and second being the namespace", url.Host)
+		}
+		serviceName := splittedHost[0]
+		namespace := splittedHost[1]
+		podPort := url.Port()
+		// infer port from service scheme when not set explicitly
+		if podPort == "" {
+			if url.Scheme == secureHTTPScheme {
+				podPort = "443"
+			} else {
+				podPort = "80"
+			}
+		}
+		// get service serviceEndpoints
+		serviceEndpoints := &corev1.Endpoints{}
+
+		err := s.kubeClient.Get(ctx, client.ObjectKey{
+			Namespace: namespace,
+			Name:      serviceName,
+		}, serviceEndpoints)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, subset := range serviceEndpoints.Subsets {
+			foundPort := ""
+			for _, port := range subset.Ports {
+				if strconv.Itoa(int(port.Port)) == podPort {
+					foundPort = fmt.Sprintf(":%d", port.Port)
+					break
+				}
+			}
+			if foundPort == "" {
+				s.logger.Info(fmt.Sprintf("Warning : could not find port %s in endpoint slice for service %s.%s definition. Will infer port from %s scheme", podPort, serviceName, namespace, url.Scheme))
+			}
+			for _, address := range subset.Addresses {
+				if address.NodeName != nil {
+					endpointUrls = append(endpointUrls, fmt.Sprintf("%s://%s%s%s", url.Scheme, address.IP, foundPort, url.Path))
+				}
+			}
+		}
+	}
+	return endpointUrls, err
+}
+
 func (s *metricsAPIScaler) getMetricValue(ctx context.Context) (float64, error) {
-	request, err := getMetricAPIServerRequest(ctx, s.metadata)
+	// if we wish to aggregate metric from a kubernetes service then we need to query each endpoint behind the service
+	if s.metadata.aggregateFromKubeServiceEndpoints {
+		endpointsUrls, err := s.getEndpointsUrlsFromServiceURL(ctx, s.metadata.url)
+		if err != nil {
+			return 0, fmt.Errorf("failed to get kubernetes endpoints urls from configured service URL")
+		}
+		if len(endpointsUrls) == 0 {
+			return 0, fmt.Errorf("no endpoints URLs were given for the service name")
+		}
+		return s.aggregateMetricsFromMultipleEndpoints(ctx, endpointsUrls)
+	}
+	// get single/unaggregated metric
+	metric, err := s.getMetricValueFromURL(ctx, nil)
+	if err == nil {
+		s.logger.V(1).Info(fmt.Sprintf("fetched single metric from metrics API url : %s. Value is %v\n", s.metadata.url, metric))
+	}
+	return metric, err
+}
+
+func (s *metricsAPIScaler) aggregateMetricsFromMultipleEndpoints(ctx context.Context, endpointsUrls []string) (float64, error) {
+	// call s.getMetricValueFromURL() for each endpointsUrls in parallel goroutines (maximum 5 at a time) and sum them up
+	const maxGoroutines = 5
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	sem := make(chan struct{}, maxGoroutines)
```

Comment:
Let's use a semaphore directly instead of a channel -> https://pkg.go.dev/golang.org/x/sync/semaphore

semaphore is part of standard lib